### ISSUE TITLE
New feature available in firmware version 1.17

### ISSUE THREAD

**Issue by robinostlund** [dannerph/homeassistant-keba#18]
Hi 😄 
I noticed today that keba released version 1.17 and noticed this feature:
```
Phase Switching via UDP enabled
For a direct initiation of the phase switching feature by an external energy management system, the KeContact P30 functionality has been enhanced to decide on itself if it should charge with 1-phase or 3-phases based on the UDP control method.
Up to now, the KeContact P30 forwarded information to the external phase switch S10, which reacted upon the command by the inverter.
```

For me it could be interesting to add to this component as it would allow me to charge my car with solar when the production is low if i could switch to one phase charging 😄

**Comment by dannerph** [comment]
Interesting, I need to check if this firmware is also ported to my BMW Wallbox Plus. However, I do not believe so ...

**Comment by dannerph** [comment]
It seems this firmware is only for the keba p30 x-series, which I do not have. Thus, I cannot test the feature by myself. However, if someone of you is capable of running some udp commands or small python scripts, we can try it.

But first of all, I would like to know, how to identify which device is capable of this feature. Could post your product code here please? (Similar to KC-P30-ES240030-000-xx)

**Comment by deqw** [comment]
My firmware is at version 1.14 so have to wait for 1.17 anyway. I assume phase switching wil be done wile not charging.

**Comment by schmersgahoven** [comment]
Hi,
yes S10 or selfmade S10 ist necessary. You ca find some by searching for Keba Phasenumschaltung. I really not have any clue of Python but was already searching for the code in the custom component folder where the UDP commands are send, with no luck. If someone tells me how I have to change files in custom component folder to have service or button to send command X2 to trigger X2 I will test. Have Keba P30 with 1.17 an can check If X2 will switch. 
You can use phase Switchring also with Keba C Series. In X Series you can activate via Front End, in C Series you have to activate by Dip switches.

**Comment by deqw** [comment]
Ordered one S10, but SolarEdge FW still on 1.14.

**Comment by dannerph** [comment]
@deqw It seems you have a Keba P30x then (or at least a branded one by SolarEdge). I have a BMW Wallbox Plus for which it is not yet clear, if the X2 can be used for the S10, as it is configured as RS485/ModBus-RTU interface. There is some discussion going on at [EVCC (German only)] (https://github.com/evcc-io/evcc/discussions/9347#discussioncomment-7110585). If I have some time, I will test to flash the newest Keba firmware to my BWM Wallbox to see if I can use the X2 or not.

I also read, that the Keba P30x is cabable to control the charging speed on its own when served with 3-phase measurement data whereas the Keba P30c only "forwards" the udp/tcp command to activate phase switching to the X2 connector. The only logic seems to be the short interuption of the charging process and blocking the next switching for a certain time. But that would not be a problem as we can add the logic in HA. I already integrated a function to set charging power instead of charging current in the python library by reading the local voltage and current values (only for charging stations with energy meter integrated).

**Comment by deqw** [comment]
@dannerph Yes correct the SolarEdge Keba P30x. Be aware that if you flash the Keba firmware into your BMW Wallbox you might loose functionality like in my case I, with the SolarEdge version of this box, will loose Eg. the Connection to the OCPP backend that is defined on the P30 Status page. Also the OCPP certificates go wrong I expect.

**Comment by deqw** [comment]
The S10 was installed yesterday and connected to X2. Also on the charger upper switch block switch 2 set ON. But still running the SolarEdge firmware 1.14 so not of any use by now. I'm considering flashing the latest Keba P30x firmware 1.17 to enable 1-phase/3-phase switching. Can I go back to SolarEdge 1.14 if needed for whatever reason?

**Comment by PulsarFX** [comment]
> The S10 was installed yesterday and connected to X2. Also on the charger upper switch block switch 2 set ON. But still running the SolarEdge firmware 1.14 so not of any use by now. I'm considering flashing the latest Keba P30x firmware 1.17 to enable 1-phase/3-phase switching. Can I go back to SolarEdge 1.14 if needed for whatever reason?
> 
> @dannerph
> 
> > There is some discussion going on at [EVCC (German only)] ([evcc-io/evcc#9347 (comment)](https://github.com/evcc-io/evcc/discussions/9347#discussioncomment-7110585)).
> 
> What I see in the documentation the communication channel for X2 can be set by user. https://www.keba.com/download/x/dea7ae6b84/kecontactp30modbustcp_pgen.pdf

You have to update one version at a time. The update package from keba download site has a pdf which states which base version is supported by the update. For me, it was sitting trough 1.14 -> 1.15.1 -> 1.16.1 and finally 1.17.1

**Comment by PulsarFX** [comment]
We only have pure keba devices. I added it to the edit, that I don't know if your edition can be updated with regular keba firmware. Sorry.

**Comment by PulsarFX** [comment]
Yeah, strange, it's missing in the dropdown
[KeContact_P30x_SW_1.15.1.zip](https://github.com/dannerph/homeassistant-keba/files/13535992/KeContact_P30x_SW_1.15.1.zip)

**Comment by deqw** [comment]
@PulsarFX Thanks for the file; got it.

Stil confused with the dip-switch settings:

The web-interface says:
![Screenshot 2023-12-02 at 11 57 35](https://github.com/dannerph/homeassistant-keba/assets/65849349/d03e91c6-caf6-42fa-a7b2-fc034dfb1f3d)

The S10 manual says:
![Screenshot 2023-12-02 at 11 56 53](https://github.com/dannerph/homeassistant-keba/assets/65849349/529d5b3f-7407-482b-a28a-340fb0ec79ce)

What description does your 1.17.1 firmware indicate via the web-interface?

**Comment by deqw** [comment]
> @deqw You have to update one version at a time. The update package from keba download site has a pdf which states which base version is supported by the update. For me, it was sitting trough 1.14 -> 1.15.1 -> 1.16.1 and finally 1.17.1
> 
> edit: I don't know if your frankenkeba would work with the actual keba firmware.

I did update to 1.17.1 finally now and all (the frankenkeba) is still working including the connection tot the SolarEdge backend.

Using the P30 web interface Phase Switching now can be enabled. Still not clear about the DSW 1.2 setting ON or OFF so asked Keba again today. But all is still useless without any status info and commands to toggle or set phase switching.

>  If I have some time, I will test to flash the newest Keba firmware to my BWM Wallbox to see if I can use the X2 or not.

 @dannerph Were you able to test this yourself with the latest firmware?

**Comment by deqw** [comment]
@dannerph Would you be so kind to add the Keba "X2 phaseSwitch" key to the integration for a new sensor to be able to switch 1 and 3 phase charging?

I tested switching by sending UDP "x2" with value 0 (1 phase) and 1  (3 phase) which works fine. Between each switch one has to wait 5 minutes otherwise "TCH-ERR:X2 : not allowed now" is returned. When the switch is applied "TCH-OK :done" is returned.

Thanks.



`{
"ID": "2",
"State": 1,
"Error1": 0,
"Error2": 0,
"Plug": 3,
"AuthON": 0,
"Authreq": 0,
"Enable sys": 1,
"Enable user": 1,
"Max curr": 0,
"Max curr %": 1000,
"Curr HW": 16000,
"Curr user": 16000,
"Curr FS": 0,
"Tmo FS": 0,
"Curr timer": 0,
"Tmo CT": 0,
"Setenergy": 0,
"Output": 1,
"Input": 0,
"X2 phaseSwitch source": 4,
"X2 phaseSwitch": 1,
"Serial": "",
"Sec": 8031
}
`

**Comment by IZZE2000** [comment]
Hi all, where we (read you) able to get the 1-phase or 3-phases based via UDP to work out via the beta? I do also have solar panels that i would like to utilize more :-)
Cheers

**Comment by dannerph** [comment]
Hi all,
I prepared the underlying library to support sending "x2" and "x2src" commands.
How should those commands be integrated into the Home Assistant integration? According to the UDP programming guide, there should be a 5 min "cool down" after triggering the swtich. I guess that is something the integration (or library should take car of?) How about setting the source? Should this be done during setup, in configuration or as service?

**Comment by robinostlund** [comment]
How about a service and a binary sensor showing if it is enabled or not?

**Comment by dannerph** [comment]
fixed in latest version 4.3.0
